DOC-9850 Document the limitation of server group data placement with respect to replica count in 7.1

RayOffiah opened the pull request:
Moved changes from https://github.com/couchbase/docs-server/pull/2300 which were attached to the incorrect branch.

RayOffiah committed:
DOC-9850 Document the limitation of server group data placement with respect to replica count in 7.1

dave-finlay commented on modules/learn/pages/clusters-and-availability/groups.adoc:
```diff
@@ -59,16 +59,35 @@ Note that Group 2 contains all the replica vBuckets that correspond to active vB
+
+*rack-zone*:: all replicas of a given vbucket must reside in separate server groups.
+
+*active balance*:: there should be approximately{empty}footnote:approximate-vbuckets[Note that by "approximately the same" we mean the number should vary by 5 vbuckets or less] the same number of active vbuckets on each node in the cluster.
```
I think we shouldn't specify a hard number on the max difference of active vbucket counts on different nodes. Generally, the active vbucket counts will only vary by 1 across nodes, but we might loosen that in the future. Also I don't want customers with 100 nodes with an average of 10 vbuckets per node complaining that some nodes might have as many as 15 vbuckets and carry 50% greater load. I'd prefer to just say the the counts are approximately the same and leave it at that.

dave-finlay commented on modules/learn/pages/clusters-and-availability/groups.adoc:
```diff
@@ -59,16 +59,35 @@ Note that Group 2 contains all the replica vBuckets that correspond to active vB
+
+*active balance*:: there should be approximately{empty}footnote:approximate-vbuckets[Note that by "approximately the same" we mean the number should vary by 5 vbuckets or less] the same number of active vbuckets on each node in the cluster.
+
+*replica balance*:: there should be approximately{empty}footnote:approximate-vbuckets[] the same number of replica vbuckets on each node in the cluster.
```
Same comment here about approximate counts of vbuckets.

dave-finlay commented on modules/learn/pages/clusters-and-availability/groups.adoc:
```diff
@@ -59,16 +59,35 @@ Note that Group 2 contains all the replica vBuckets that correspond to active vB
+
+*replica balance*:: there should be approximately{empty}footnote:approximate-vbuckets[] the same number of replica vbuckets on each node in the cluster.
+
+Not all the constraints can be satisfied when the buckets are allocated across  uneven group. 
```
extra space

RayOffiah replied:
Good spot, though Asciidoc removes double spaces like this when the page is compiled.

dave-finlay commented on modules/learn/pages/clusters-and-availability/groups.adoc:
```diff
@@ -59,16 +59,35 @@ Note that Group 2 contains all the replica vBuckets that correspond to active vB
+
+Not all the constraints can be satisfied when the buckets are allocated across  uneven group. 
+In this scenario, the `active balance` and `rack-zone` constraints will take priority: 
+when the vbucket map is generated, the aim is to ensure that there are the same number of active vbuckets on each cluster and that replicas of a given vbucket must reside in separate groups.
```
suggest "we will" instead of "the aim is"

dave-finlay replied:
approximately the same

dave-finlay commented on modules/learn/pages/clusters-and-availability/groups.adoc:
```diff
@@ -59,16 +59,35 @@ Note that Group 2 contains all the replica vBuckets that correspond to active vB
-However, since the groups contain unequal number of nodes, Group 2 not only contains all the replica vBuckets that correspond to active vBuckets on Group 1, but also contains all the replica vBuckets for its own additional node, Server 9 — the replicas for Server 9 being distributed across the other Group 2 nodes; which are Servers 5, 6, 7, and 8.
-Server 9 contains its own active vBuckets, plus replica vBuckets for Group 1.
+NOTE: Since smaller server groups will carry more replicas on each node, their performance may not be the same as the nodes in the bigger server groups. 
+For best performance, try to maintain equality of server groups.
```
We need to make this recommendation more strongly. Smaller server groups will carry more replica vbuckets, which means there is greater memory pressure on memcached so more fetches go to disk which means higher worst case GET latencies. Additionally more replicas mean more writes to disk and greater compaction burden which will also affect latencies. Customers will notice this as the smaller server groups will "perform" worse than the later server groups. So for reasons of consistency of performance Couchbase strong recommends that customers balance their server groups.

dave-finlay reviewed (commented):
initial set of comments left

RayOffiah committed:
Fixes following Dave Finlay's review

RayOffiah commented:
@dave-finlay <-- Fixes pushed.

RayOffiah committed:
Drop shadows missing off the diagrams.

RayOffiah committed:
Added note detailing the new arrangement for replicas.

RayOffiah commented:
To make it easier to read, I've put up the page here:

https://rayoffiah.github.io/DOC-9850/server/7.1/learn/clusters-and-availability/groups.html

tonyjhillman reviewed (approved):
Line 85 'group' should be plural.

The spelling of vBucket varies, so should probably be made consistent.

Line 104 should probably be 'a greater compaction burden'.

Line 107 'strong' should be 'strongly'.

The statement that customers should 'balance' their server groups I think means that there should be the same number of nodes in each. But they might interpret it as 'rebalance', so I wonder if another phrase would be better.

RayOffiah committed:
Few corrections following Tony's review

dave-finlay commented on modules/learn/pages/clusters-and-availability/groups.adoc:
```diff
@@ -46,6 +47,18 @@ For information on the standard (non-Group-based) distribution of replica vBucke
 The distribution of vBuckets across groups is exemplified below.
 In each illustration, all servers are assumed to be running the Data Service.
 
+The number of replicas of each bucket in each group depends on the number of server groups in the rack.
```
It would be good if we could do the following:

- It would be to highlight this as "Important Note" or "Pay Attention Here" -- however we generally do that in the docs.
- We should also tell the users that this is a behavior change from 7.0 and prior releases. 
- We should emphasize that this behavior change only affects situations where the user has configured non-trivial server groups (i.e. more than one server group with that contain KV nodes).

dave-finlay replied:
Instead of the "The number of replicas of each bucket", we should say: "the number of replicas placed in server groups depends on the number of configured server groups." 

May be a good idea to introduce the idea of "non-trivial KV server groups" here. Non-trivial is a terrible name but it's the best I can come up with with little thought. Perhaps Shivani might have a better name. Worth getting a review from her, I would say.

RayOffiah committed:
Expanding note on changes from 7.0x into an admonition block.

RayOffiah commented:
Added the details in an admonition

<img width="834" alt="image" src="https://user-images.githubusercontent.com/77050471/160804174-ef76d78f-15cc-49f4-a955-af4f1e2efcac.png">

@shivaniguptacb <-- How does this look?

dave-finlay commented on modules/learn/pages/clusters-and-availability/groups.adoc:
```diff
@@ -46,6 +47,23 @@ For information on the standard (non-Group-based) distribution of replica vBucke
 
+[IMPORTANT]
+====
+The number of replicas placed in server groups depends on the number of configured server groups..
```
double period

dave-finlay commented on modules/learn/pages/clusters-and-availability/groups.adoc:
```diff
@@ -46,6 +47,23 @@ For information on the standard (non-Group-based) distribution of replica vBucke
+[IMPORTANT]
+====
+The number of replicas placed in server groups depends on the number of configured server groups..
+If there are at least two server groups, then we can broadly say that:
```
Suggest: "In 7.1, when there are at least two server groups containing KV nodes, then it is the case that"

dave-finlay commented on modules/learn/pages/clusters-and-availability/groups.adoc:
```diff
@@ -46,6 +47,23 @@ For information on the standard (non-Group-based) distribution of replica vBucke
+
+[stem]
+++++
+number\:of\:replicas = number\:of\:server\:groups - 1
```
Suggest "number of replicas placed" instead of just "number of replicas"

RayOffiah committed:
Removed extra full stop.

shivaniguptacb commented on modules/learn/pages/clusters-and-availability/groups.adoc:
```diff
@@ -46,6 +47,23 @@ For information on the standard (non-Group-based) distribution of replica vBucke
 The distribution of vBuckets across groups is exemplified below.
 In each illustration, all servers are assumed to be running the Data Service.
 
+[IMPORTANT]
```
This whole section 'Important' is  super complicated. I suggest we simplify it to something as follows. Also place it after the discussion on equal and unequal groups (not right at the top of the section):

If you have fewer server groups than the total number of data copies (= 1 + number of replicas), only as many copies will be placed as the number of server groups. Essentially you may not get as many replicas as you set if you have fewer server groups. 
E.g. if you have 2 server groups, and set replicas = 2, you will only have one replica for each vbucket. But if you have 3 server groups you will have 2 replicas for each vbucket.
This behavior is a change from pre-7.1 behavior where we did not reduce the number of data copies to match the number of server groups.

RayOffiah committed:
More changes following review: Made thew warning a bit more prominent, so it doesn't get lost around the other warnings.
Reworded the warning for clarity.
Moved it to the bottom of the section of equal/unequal groups.

RayOffiah committed:
Changed vbucket to vBucket

RayOffiah commented:
https://rayoffiah.github.io/DOC-9850/server/7.1/learn/clusters-and-availability/groups.html

<img width="833" alt="image" src="https://user-images.githubusercontent.com/77050471/161017368-b0d6d48a-ca56-403b-b3d3-caaf0280666a.png">

dave-finlay commented on modules/learn/pages/clusters-and-availability/groups.adoc:
```diff
@@ -59,16 +60,60 @@ Note that Group 2 contains all the replica vBuckets that correspond to active vB
+
+For more information on optimizing your cluster configuration, consult the xref:install:sizing-general.adoc[].
+
+.IMPORTANT - Change in behavior from previous versions
```
Looks good to me.

shivaniguptacb commented on modules/learn/pages/clusters-and-availability/groups.adoc:
```diff
@@ -1,6 +1,7 @@
 = Server Group Awareness
```
There is a para here that should be removed as it is no longer true. It is the following one:

"Protection is optimal when groups are assigned equal numbers of nodes, and replica vBuckets and index replicas are thereby distributed such that none ever occupies the same group as its associated active vBucket or index. By contrast, when groups are not assigned equal numbers of nodes, rebalance can only produce a best effort redistribution of replicas: this may result in one or more replica vBuckets and/or index replicas each occupying the same group as their associated active vBuckets or indexes; meaning that vBucket data and/or index data may be lost if such a group becomes unavailable."

shivaniguptacb replied:
Also the following bullet should be removed:
"Failover should be enabled for server groups only if three or more server groups have been established, and sufficient capacity exists to absorb the load of any failed-over group."

shivaniguptacb replied:
This bullet also should be removed:
"If a cluster contains only one node, or only one group of nodes, enabling Group Failover has no effect; meaning that failover can only be performed on a per-node basis."

shivaniguptacb replied:
Reword "For information on failing over individual nodes, see [Fail a Node over and Rebalance]" to 
"For information on failover see [Failover and Rebalance]" with same link

RayOffiah committed:
Corrections

shivaniguptacb reviewed (approved):
Looks good now.
